Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1918,66 +1937,134 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
+            (fabs(finalTextTransform.d * glyphRuns[i].run->fontEmSize) <= c_glyphThreshold &&
+             fabs(finalTextTransform.c * glyphRuns[i].run->fontEmSize) <= c_glyphThreshold)) {
+            TraceWarning(TAG, L"Glyphs too small to be rendered");
+
+            // Not a failure state! Not drawing the glyphs is *okay*.
+            return S_FALSE;
+        }
+    }
 
-        // Not a failure state! Not drawing the glyphs is *okay*.
-        return S_FALSE;
+    // Get the linear transformation that is the inverse of the text transform
+    // We transform the positions of each glyph by the inverse so they will be positioned correctly
+    // While still being transformed per glyph by the text transformation
+    CGAffineTransform invertedTextTransformation = CGAffineTransformInvert(
+        CGAffineTransformScale(CGAffineTransformMake(textTransform.a, textTransform.b, textTransform.c, textTransform.d, 0, 0), 1, -1));
+
+    // Vectors which hold the positions of each glyph that we give to DWRITE for drawing
+    std::vector<std::vector<DWRITE_GLYPH_OFFSET>> positionsVectors(runsCount);
+
+    // Used for creating vector of advances to be shared
+    size_t maxGlyphCount = 0;
+
+    for (size_t i = 0; i < runsCount; ++i) {
+        positionsVectors[i].resize(glyphRuns[i].run->glyphCount);
+        maxGlyphCount = std::max(maxGlyphCount, glyphRuns[i].run->glyphCount);
+        if (transformByGlyph) {
+            // First glyph's origin is at the given relative position for the glyph run
+            CGPoint runningPosition{ glyphRuns[i].relativePosition.x, std::round(glyphRuns[i].relativePosition.y) };
+            DWRITE_GLYPH_OFFSET previousOffset{ 0, 0 };
+            for (size_t j = 0; j < glyphRuns[i].run->glyphCount; ++j) {
+                // Invert position by text transformation
+                CGPoint transformedPosition = CGPointApplyAffineTransform(runningPosition, invertedTextTransformation);
+
+                // Set current glyph's position to be the actual position transformed by the inverted text position
+                // So when the space is transformed by the text position it will be drawn in the correct real position
+                positionsVectors[i][j] =
+                    DWRITE_GLYPH_OFFSET{ transformedPosition.x + glyphRuns[i].run->glyphOffsets[j].advanceOffset,
+                                         std::round(transformedPosition.y + glyphRuns[i].run->glyphOffsets[j].ascenderOffset) };
+
+                // Translate position of next glyph by current glyph's advance
+                runningPosition.x += glyphRuns[i].run->glyphAdvances[j];
+            }
+        } else {
+            // CG text drawing mathods use the text matrix to transform about the origin, not about each glyph's origin
+            // We can simply use the positions directly from the glyph run
+            std::copy(glyphRuns[i].run->glyphOffsets,
```

Comment:
if you avoided the copy, you could also avoid the resize up top :0

---
In reply to: [100875602](https://github.com/Microsoft/WinObjC/pull/1933#discussion_r100875602) [](ancestors = 100875602)